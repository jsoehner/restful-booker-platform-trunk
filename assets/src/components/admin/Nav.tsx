import React, { useState } from 'react';
import Link from 'next/link';
import { usePathname } from 'next/navigation';
import Cookies from 'universal-cookie';

interface NavProps {
  setAuthenticate: (value: boolean) => void;
  isAuthenticated: boolean | null;
}

const Nav: React.FC<NavProps> = ({ setAuthenticate, isAuthenticated }) => {
  const pathname = usePathname();
  const [messageCount, setMessageCount] = useState(0);

  const fetchMessageCount = async () => {
    try {
      const response = await fetch('/api/message/count', {
        cache: 'no-store',
        headers: {
          'Pragma': 'no-cache',
          'Cache-Control': 'no-cache'
        }
      });
      if (response.ok) {
        const data = await response.json();
        setMessageCount(data.count || 0);
      }
    } catch (error) {
      console.error('Error fetching message count:', error);
    }
  };

  React.useEffect(() => {
    if (isAuthenticated) {
      fetchMessageCount();
    }
  }, [isAuthenticated]);

  // Expose the fetchMessageCount function globally
  React.useEffect(() => {
    if (typeof window !== 'undefined') {
      (window as any).updateMessageCount = fetchMessageCount;
    }
  }, []);

  const doLogout = () => {
    const cookies = new Cookies();
    cookies.remove('token', { path: '/' });
    setAuthenticate(false);
    window.location.href = '/';
  };

  const getNavLinkClass = (path: string) => {
    return pathname === path ? 'nav-link active' : 'nav-link';
  };

  return (
    <nav className="navbar navbar-expand-md navbar-dark bg-dark mb-4">
      <div className="container-fluid">
        <Link href="/" className="navbar-brand">
          Restful Booker Platform Demo
        </Link>
        <button
          className="navbar-toggler"
          type="button"
          data-toggle="collapse"
          data-target="#navbarSupportedContent"
          aria-controls="navbarSupportedContent"
          aria-expanded="false"
          aria-label="Toggle navigation"
        >
          <span className="navbar-toggler-icon"></span>
        </button>
        <div className="collapse navbar-collapse" id="navbarSupportedContent">
          <ul className="navbar-nav mr-auto">
            {isAuthenticated && (
              <>
                <li className="nav-item">
                  <Link href="/admin/rooms" className={getNavLinkClass('/admin/rooms')}>
                    Rooms
                  </Link>
                </li>
                <li className="nav-item">
                  <Link id="reportLink" href="/admin/report" className={getNavLinkClass('/admin/report')}>
                    Report
                  </Link>
                </li>
                <li className="nav-item">
                  <Link id="brandingLink" href="/admin/branding" className={getNavLinkClass('/admin/branding')}>
                    Branding
                  </Link>
                </li>
                <li className="nav-item">
                  <Link href="/admin/message" className={getNavLinkClass('/admin/message')}>
                    Messages {messageCount > 0 && <span className="badge bg-danger text-white">{messageCount}</span>}
                  </Link>
                </li>
              </>
            )}
          </ul>
          <ul className="navbar-nav ms-auto">
            <li className="nav-item">
              <a className="nav-link" id="frontPageLink" href="/">Front Page</a>
            </li>
            <li className="nav-item">
              <button onClick={doLogout} className="btn btn-outline-danger my-2 my-sm-0">
                Logout
              </button>
            </li>
          </ul>
        </div>
      </div>
    </nav>
  );
};

export default Nav; 