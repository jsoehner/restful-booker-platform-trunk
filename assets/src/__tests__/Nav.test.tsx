import React from 'react';
import Nav from '../components/admin/Nav';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

const mockUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockUseNavigate,
}));

describe('Nav Component', () => {
  test('Nav bar renders', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <Nav setAuthenticate={() => {}} isAuthenticated={true} />
      </BrowserRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
