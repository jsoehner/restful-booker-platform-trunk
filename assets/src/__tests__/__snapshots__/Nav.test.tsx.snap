// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav Component Nav bar renders 1`] = `
<DocumentFragment>
  <nav
    class="navbar navbar-expand-md navbar-dark bg-dark mb-4"
  >
    <div
      class="container-fluid"
    >
      <a
        class="navbar-brand"
        href="/"
      >
        Restful Booker Platform Demo
      </a>
      <button
        aria-controls="navbarSupportedContent"
        aria-expanded="false"
        aria-label="Toggle navigation"
        class="navbar-toggler"
        data-target="#navbarSupportedContent"
        data-toggle="collapse"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="collapse navbar-collapse"
        id="navbarSupportedContent"
      >
        <ul
          class="navbar-nav mr-auto"
        >
          <li
            class="nav-item"
          >
            <a
              class="nav-link"
              href="/admin/rooms"
            >
              Rooms
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link"
              href="/admin/report"
              id="reportLink"
            >
              Report
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link"
              href="/admin/branding"
              id="brandingLink"
            >
              Branding
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link"
              href="/admin/message"
            >
              Messages 
            </a>
          </li>
        </ul>
        <ul
          class="navbar-nav ms-auto"
        >
          <li
            class="nav-item"
          >
            <a
              class="nav-link"
              href="/"
              id="frontPageLink"
            >
              Front Page
            </a>
          </li>
          <li
            class="nav-item"
          >
            <button
              class="btn btn-outline-danger my-2 my-sm-0"
            >
              Logout
            </button>
          </li>
        </ul>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;
